### PR TITLE
[wip] Fix some bugs for TPU [skip ci]

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/tensorboard.py [hunk 1]
```diff
@@ -198,7 +198,9 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
                 self.experiment.add_scalars(k, v, step)
             else:
                 try:
+                    print("before", k, v, step)
```
Comment (lezwon) [inline pytorch_lightning/loggers/tensorboard.py:201]: Guess we could remove this.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -700,12 +700,18 @@ def run_evaluation(self, max_batches=None, on_epoch=False):
...
+            print("dl_outputs")
+
         if self._predicting:
             return self.evaluation_loop.on_predict_epoch_end()
 
         # lightning module method
         deprecated_eval_results = self.evaluation_loop.evaluation_epoch_end()
 
+        print(self.current_epoch)
+
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/trainer.py:712]: Could remove the changes in this file.

File: pytorch_lightning/loggers/tensorboard.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loggers/tensorboard.py:201]: Still debugging :)

Comment (tchaton) [reply pytorch_lightning/loggers/tensorboard.py]: Done !

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 3]
```diff
@@ -373,7 +373,10 @@ def select_training_type_plugin(self):
         elif self.use_horovod:
             plugin = HorovodPlugin(parallel_devices=self.parallel_devices)
         elif self.on_tpu:
-            plugin = SingleTPUPlugin(self.tpu_id)
+            if isinstance(self.tpu_cores, list):
+                plugin = SingleTPUPlugin(self.tpu_id)
+            else:
+                plugin = TPUSpawnPlugin(parallel_devices=list(range(self.tpu_cores)))
```
Comment (lezwon) [inline pytorch_lightning/accelerators/accelerator_connector.py:379]: this was just a temporary fix. I think `TPUSpawnPlugin` is meant to be called above within `elif self.use_ddp` block.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -67,14 +80,15 @@ def new_process(self, process_idx: int, trainer) -> None:
...
+        trainer.accelerator_backend.setup_optimizers(trainer)
         self.barrier()
 
         if trainer.testing:
             results = trainer.run_test()
         else:
             results = trainer.train()
 
-        self.__save_end_of_training_weights(self.lightning_module)
+        self.__save_end_of_training_weights(self.lightning_module, trainer)
```
Comment (lezwon) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:91]: dont think `trainer ` arg is used. Maybe we can omit it from the method signature itself.

**Review by lezwon — APPROVED** [review]
LGTM

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -197,6 +197,7 @@ def on_pretrain_routine_start(self, trainer, pl_module):
         self.__resolve_ckpt_dir(trainer)
         self.save_function = trainer.save_checkpoint
 
+    @rank_zero_only
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:200]: I don't think this is correct. See the `save_checkpoint()` docstring

File: pytorch_lightning/utilities/seed.py [hunk 6]
```diff
@@ -51,7 +50,7 @@ def seed_everything(seed: Optional[int] = None) -> int:
         rank_zero_warn(f"{seed} is not in bounds, numpy accepts from {min_seed_value} to {max_seed_value}")
         seed = _select_seed_randomly(min_seed_value, max_seed_value)
 
-    log.info(f"Global seed set to {seed}")
+    rank_zero_info(f"Global seed set to {seed}")
```
Comment (carmocca) [inline pytorch_lightning/utilities/seed.py:53]: We chose to log on all ranks so users can be sure the same seed is used in all processes.

Comment (tchaton) [reply pytorch_lightning/utilities/seed.py:53]: Makes sense.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 7]
```diff
@@ -400,11 +400,11 @@ def save_checkpoint(self, filepath, weights_only: bool = False):
         """
         # dump states as a checkpoint dictionary object
         checkpoint = self.dump_checkpoint(weights_only)
-
         if self.trainer.is_global_zero:
             # write the checkpoint dictionary on the file
-            if self.trainer.accelerator_backend:
-                checkpoint = self.trainer.accelerator_backend.on_save(checkpoint)
+
+            if self.trainer.training_type_plugin:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:406]: trainer/accelerator always has a traintype plugin, so we could even drop the if condition :)

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 7 above)
Comment (lezwon) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:406]: I think during sanity check the `accelerator_backend` was not set. Not sure if that's the case now.

**Pull request merged**